### PR TITLE
refactor: cursor-based pagination for comments

### PR DESCRIPTION
**Pull request by iuioiua**
This change introduces a pagination style that's more natural for Deno KV. The kind of pagination is unconventional but remarkably simple, IMO. Once merged, we'll make this the default pagination method for the site.

Note: the last page is empty. This occurs on the last iteration of the async iterator when the [`limit` option](https://deno.land/api@v1.36.1?s=Deno.KvListOptions&unstable=#prop_limit) is divisible by the total number of results of the async iterator. E.g. `limit = 2` and total results = 10, as in this video. Is this a bug?

https://github.com/denoland/saaskit/assets/29347852/d3e144e2-25db-4e5a-82ba-fa4073677850

Towards #414

### CODE REVIEW

**Comment by kt3k** [comment]
> This occurs on the last iteration of the async iterator when the [limit option](https://deno.land/api@v1.36.1?s=Deno.KvListOptions&unstable=#prop_limit) is divisible by the total number of results of the async iterator. E.g. limit = 2 and total results = 10, as in this video. Is this a bug?

I think one of the common ways to avoid that behavior (probably a bug) is to always get `(number of visible items) + 1` to detect if the next item exist or not.

**Comment by iuioiua** [comment]
> I think one of the common ways to avoid that behavior (probably a bug) is to always get `(number of visible items) + 1` to detect if the next item exist or not.

Nice suggestion! I've added a workaround and raised an issue. I'd like to hear some thoughts on the feel of this new type of pagination. @brunocorrea23 and @mbhrznr, WDYT?

File: routes/items/[id].tsx [hunk 1]
```diff
@@ -91,23 +88,23 @@ export default async function ItemsItemPage(
...
-  const allComments = await getCommentsByItem(id);
-  const comments = allComments
-    .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime())
-    .slice((pageNum - 1) * PAGE_LENGTH, pageNum * PAGE_LENGTH);
+  const cursor = ctx.url.searchParams.get("cursor") ?? undefined;
+  const iter = listCommentsByItem(itemId, { cursor });
+  const comments = await valuesFromIter(iter);
+
+  /** @todo https://github.com/denoland/deno/issues/20173 */
+  const nextPageIter = listCommentsByItem(itemId, { cursor: iter.cursor });
```
Comment (losfair) [inline routes/items/[id].tsx:100]: Do not do a second list. Instead, increase the limit by one in the first `listCommentsByItem` call and drop the last entry.

Comment (iuioiua) [reply routes/items/[id].tsx]: Done. Is this solution optimal? WDYT?

**Comment by mbhrznr** [comment]
had the chance to checkout and test the feature.
the usage feels nice and the transitions were quick, yet i was wondering if this is the best approach ux wise.

this way the pagination would be uni-directional and there would be no "easy access" to specific comments.
while it would be easy to show oldest comments first via `reverse`, the "in-between" might be harder to access.
and a user would have to start over to (re-)read any comments.

with the `cursor` search param, the url would be less predictable. 
while negligible, it'd make manual url updates almost impossible (when compared to a more "traditional" pagination w/ `page=<number>`.
however as this approach would still provide sharable urls, i'd consider it insignificant.

as always, great work! 👏 👏

**Review by losfair — APPROVED** [review]
LGTM for the pagination logic

**Comment by iuioiua** [comment]
Closing in favour of #429 or #438

**Pull request closed without merging**